pagespeed/system/controller_manager.h

```diff
@@ -0,0 +1,141 @@
+#define PAGESPEED_CONTROLLER_MANAGER_H_
+
+#include <sys/types.h>
+
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/file_system.h"
+#include "pagespeed/kernel/base/message_handler.h"
+#include "pagespeed/kernel/base/scoped_ptr.h"
+#include "pagespeed/kernel/base/string.h"
+#include "pagespeed/kernel/base/thread.h"
+#include "pagespeed/kernel/base/thread_system.h"
+#include "pagespeed/kernel/base/timer.h"
+#include "pagespeed/system/system_rewrite_driver_factory.h"
+
+
+namespace net_instaweb {
+
+// A server context with features specific to a psol port on a unix system.
+class ControllerManager {
+ public:
+  // Takes ownership of nothing.
+  ControllerManager(SystemRewriteDriverFactory* factory,
+                    ThreadSystem* thread_system,
+                    FileSystem* file_system,
+                    Timer* timer,
+                    MessageHandler* handler);
+  virtual ~ControllerManager() {}
+
+  // Called on system startup, before forking off any workers.  Starts up a
+  // babysitter process that starts a controller process and restarts the
+  // controller if it dies.  Also called (again) on configuration reloading.
+  void ForkOffControllerProcess();
+
+  class ProcessDeathWatcherThread : public ThreadSystem::Thread {
+   public:
+    // Takes ownership of nothing.  Not that it matters, since we run until we
+    // exit.
+    ProcessDeathWatcherThread(ThreadSystem* thread_system,
+                              Timer* timer,
+                              MessageHandler* handler,
+                              pid_t master_pid,
+                              pid_t babysitter_pid);
+
+    virtual void Run();
+
+   private:
+    Timer* timer_;
+    MessageHandler* handler_;
+    pid_t master_pid_;
+    pid_t babysitter_pid_;
+
+    const char* controller_pid_filename_;
+
+    DISALLOW_COPY_AND_ASSIGN(ProcessDeathWatcherThread);
+  };
+
+ private:
+  // Controller will be hooked up here.  This method is called in a single
+  // centralized process, and if that process dies it will be started again.
+  void RunController();
+
+  // Remove inherited signal handlers in babysitter/controller, and replace them
+  // with our cleanups as needed.
+  void SetUpSignalHandlers();
+
+  // Set us up as a proper daemon, with no stdin/out/err and not process group.
+  void Daemonize();
+
+  // Resolves any relativity in the path as specified in the config.
+  GoogleString ControllerPidFilename();
+
+  // Returns -1 on failure, pid otherwise.
+  pid_t ReadPidFromFile(const GoogleString& filename);
+
+  // Look up the PID of the controller process.  If the controller process is
+  // not running, returns -1 immediately.
+  pid_t ControllerPid();
+
+  // TODO(jefftk): This is confusing verging on misleading.  This actually
+  // writes the babysitter pid, and the way we kill the controller is to kill
+  // the babysitter.  So the "controller pid file" actually contains the pid of
+  // the babysitter process.
```

Comment:
Yes, sorry, cleared this up some.   It's the babysitter pid in our code, and the "controller babysitter pid" to users.